### PR TITLE
[PACKAGE] Add Retroarch v1.16.0.3

### PR DESCRIPTION
**Pull request by tiopex**
Add Retroarch v1.16.0.3 Package with all cores (Latest) which are included in the official release. 
All are compiled from the source.

Should work for uclibc and musl.

### CODE REVIEW

**Comment by TriForceX** [comment]
Looks good! both worked fine to me.

File: package/retroarch/0002-use-home-dir-not-hidden.patch [hunk 1]
```diff
@@ -0,0 +1,13 @@
+diff --git a/dingux/dingux_utils.c b/dingux/dingux_utils.c
+index c6d5b791bf..7f2ca9ed92 100644
+--- a/dingux/dingux_utils.c
++++ b/dingux/dingux_utils.c
+@@ -37,7 +37,7 @@
+ /* Base path defines */
+ #define DINGUX_HOME_ENVAR                 "HOME"
+ #define DINGUX_BASE_DIR                   "retroarch"
+-#define DINGUX_BASE_DIR_HIDDEN            ".retroarch"
++#define DINGUX_BASE_DIR_HIDDEN            "retroarch"
```
Comment (Apaczer) [inline package/retroarch/0002-use-home-dir-not-hidden.patch]: I understand the limitation of genimage, but you should use some runtime method to change that back to normal - e.g. manifest or firstboot. Ideally I would see at some point an IPK package ready to be extracted by enduser with preinst shell instructions.

File: package/retroarch/0003-video-sync.patch [hunk 2]
```diff
@@ -0,0 +1,22 @@
+diff --git a/config.def.h b/config.def.h
+index a59e56441b..0c002b97b6 100644
+--- a/config.def.h
++++ b/config.def.h
+@@ -1119,7 +1119,7 @@
+ #endif
+ 
+ /* Will sync audio. (recommended) */
+-#define DEFAULT_AUDIO_SYNC true
++#define DEFAULT_AUDIO_SYNC false
```
Comment (Apaczer) [inline package/retroarch/0003-video-sync.patch]: why these changes? I had bad experience with audio sync OFF

File: package/libretro-cores/libretro-cores.mk [hunk 3]
```diff
@@ -0,0 +1,48 @@
+LIBRETRO_CORES_VERSION = origin/master
```
Comment (Apaczer) [inline package/libretro-cores/libretro-cores.mk]: This is for libretro-super. 
Is there a way to pick stable releases of libretro cores? Maybe different them between every core (pick the one most reliable)? We should inspect more `libretro-fetch.sh`

File: package/libretro-cores/0002-fix-rules.d.patch [hunk 4]
```diff
@@ -0,0 +1,22 @@
+diff --git a/rules.d/core-rules.sh b/rules.d/core-rules.sh
+index f80055be..a355d27b 100644
+--- a/rules.d/core-rules.sh
++++ b/rules.d/core-rules.sh
+@@ -1121,7 +1121,7 @@ include_core_picodrive() {
+ libretro_picodrive_name="Picodrive"
+ libretro_picodrive_git_url="https://github.com/libretro/picodrive.git"
+ libretro_picodrive_git_submodules="yes"
+-libretro_picodrive_build_makefile="Makefile.libretro"
++libretro_picodrive_build_makefile="Makefile.libretro -f ../Makefile.libretro-picodrive"
```
Comment (Apaczer) [inline package/libretro-cores/0002-fix-rules.d.patch]: Build fails the normal way? Maybe we should alert libretro folks.

**Review by Apaczer — COMMENTED** [review]
Always suprising us :+1: . 
Will give it a go for a testing in comming days, see if any core falls outside general libretro buildbot builds.

File: package/retroarch/0002-use-home-dir-not-hidden.patch (hunk 1 above)
Comment (tiopex) [reply package/retroarch/0002-use-home-dir-not-hidden.patch]: Yes, I had that option too, to use manifest, but it still workaround, but from other side, I don't see any reason why not using not hidden homedir for ra (gmenu2x also using not hidden)

File: package/retroarch/0003-video-sync.patch (hunk 2 above)
Comment (tiopex) [reply package/retroarch/0003-video-sync.patch]: After tests this configuration works better.

File: package/libretro-cores/libretro-cores.mk (hunk 3 above)
Comment (tiopex) [reply package/libretro-cores/libretro-cores.mk]: Libretro cores are not versioned, like retroarch,

File: package/libretro-cores/0002-fix-rules.d.patch (hunk 4 above)
Comment (tiopex) [reply package/libretro-cores/0002-fix-rules.d.patch]: For miyoo platform many cores have hardcoded path In var CC=/opt/miyoo/usr/bin/arm-linux-gcc

Most of them I can override using env, but in case of picodrive is tricky, because inside during compilation they reassign CC to host-gcc.

The best option for us is to have extra env with path to gcc like for other platforms e.g.
CC=$EXTRA_PATH/arm-linux-gcc

File: package/retroarch/0003-video-sync.patch (hunk 2 above)
Comment (tiopex) [reply package/retroarch/0003-video-sync.patch]: In default configuration they sync video to audio, so when you turn off the audio, there is no FPS limit, and for example quicknes runs at 200fps which cause the game is incredible fast. In this configuration sync is done to the LCD panel 60hz-60fps

File: package/retroarch/0002-use-home-dir-not-hidden.patch (hunk 1 above)
Comment (Apaczer) [reply package/retroarch/0002-use-home-dir-not-hidden.patch]: one reason would be that you should keep custom build at most inline with official ones, so people wont endup with unnecessary or double stuff. I'll take a look in gmenu2x src (not sure why smbd left it that way).

File: package/retroarch/0003-video-sync.patch (hunk 2 above)
Comment (Apaczer) [reply package/retroarch/0003-video-sync.patch]: Ok, we could push it upstream some day if that's the case.

File: package/libretro-cores/libretro-cores.mk (hunk 3 above)
Comment (Apaczer) [reply package/libretro-cores/libretro-cores.mk]: Yes that's why ppl use commit sha to pick the best option in that case. Otherwise why we do it if we have daily builds from official buildbot, we could as well wget artifact, check sha265 and extract. Aside from that, theres big bottleneck in package buildtime

File: package/retroarch/0002-use-home-dir-not-hidden.patch (hunk 1 above)
Comment (tiopex) [reply package/retroarch/0002-use-home-dir-not-hidden.patch]: ok, I will revert to hidden and use firstboot to rename dir

File: package/libretro-cores/libretro-cores.mk (hunk 3 above)
Comment (tiopex) [reply package/libretro-cores/libretro-cores.mk]: I will abandon the libretro-super repository and try to build the cores as separate packages

File: package/libretro-cores/libretro-cores.mk
Comment (Apaczer) [inline package/libretro-cores/libretro-cores.mk:1]: You could leave that package as good example to build all cores from latest commit in one go. If there's a conflict with new configuration (e.g. `core-info`) just add !BR2_PACKAGE_... in depends.

File: package/libretro-database/Config.in [hunk 5]
```diff
@@ -1,5 +1,6 @@
 config BR2_PACKAGE_LIBRETRO_DATABASE
     bool "retroarch database"
+    depends on BR2_PACKAGE_RETROARCH
```
Comment (Apaczer) [inline package/libretro-database/Config.in:3]: it would be good to add some config dependecies in `/package/retroarch` itself

File: board/miyoo/boot/firstboot [hunk 6]
```diff
@@ -599,6 +599,8 @@ if test -r "${BASEDIR}/configs/manifest"; then
 		fi
 	done
 fi
+mv "${HOME}/retroarch" "${HOME}/.retroarch" | tee -a ${LOG}
```
Comment (Apaczer) [inline board/miyoo/boot/firstboot:603]: Move it to `manifest`. It is hard to grasp what's happening within firstboot's wall of code.

File: package/Config.in [hunk 7]
```diff
@@ -291,7 +291,76 @@ menu "Games"
 	source "package/opentyrian-data/Config.in"
 	source "package/prboom/Config.in"
 	source "package/retroarch/Config.in"
-	source "package/libretro-cores/Config.in"
+	source "package/libretro-core-info/Config.in"
```
Comment (Apaczer) [inline package/Config.in:297]: Could you say, from what are you basing your work? I mean whole PR, I've found similar implementation in https://github.com/mac2612/retroleap , but wouldn't be surprised if it was labor of your hands.

**Review by Apaczer — COMMENTED** [review]
_magnifique!_ 
I think we can rid off the precompiled stuff now (like emus) from board. In the next step we should add appLinks for every core in relevant sections.

File: package/libretro-database/Config.in (hunk 5 above)
Comment (tiopex) [reply package/libretro-database/Config.in:3]: What kind of config? This dependency is only for quick unselect all cores by unselecting only retroarch

File: board/miyoo/boot/firstboot (hunk 6 above)
Comment (tiopex) [reply board/miyoo/boot/firstboot:603]: But this manifest is used to copy files from boot partition to main, not to rename directory

File: package/Config.in (hunk 7 above)
Comment (tiopex) [reply package/Config.in:297]: nah, I'm not so crazy xD, based on https://gitlab.com/recalbox/recalbox

File: board/miyoo/boot/firstboot (hunk 6 above)
Comment (Apaczer) [reply board/miyoo/boot/firstboot:603]: yes sry, we should then change to `mv` cmd or ditch `manifest` completly - it was only needed for old firstboot method. 
I'm fine with what you did, but add some echo msg for stdout output and logs about what's happening.

File: package/Config.in (hunk 7 above)
Comment (Apaczer) [reply package/Config.in:297]: ma man!

File: package/libretro-database/Config.in (hunk 5 above)
Comment (Apaczer) [reply package/libretro-database/Config.in:3]: I'm talking about Config.in for BR2_PACKAGE_RETROARCH, like add there:
depends on !BR2_STATIC_LIBS
select BR2_PACKAGE_LIBRETRO_CORE_INFO

not sure what else do we need.

**Comment by Apaczer** [comment]
There is also https://github.com/libretro-thumbnails/libretro-thumbnails , but some of submodules are a few GBs
large. We would have to downsize all *.img as soon as fetching submodules and see if there's big overhead in buildtime.

In the meantime I added icons for `retroarch` in gmenu2x.

There is small issue of using /usr/bin/ as working dir for running apps. You can't kill them with our kernel/daemon hotkeys, moreover random ppl using Windows will have an issue updating their binary. For now I think u should make entry in genimage.sh to move it to userspace, after 2.0.0 release I could change shipping method of RetroArch in upstream.

**Comment by tiopex** [comment]
> There is also https://github.com/libretro-thumbnails/libretro-thumbnails , but some of submodules are a few GBs large. We would have to downsize all *.img as soon as fetching submodules and see if there's big overhead in buildtime.

No, I checkout  via git is slow, I'm trying to use tar.gz where it is posible
> 
> In the meantime I added icons for `retroarch` in gmenu2x.

Nice!

> 
> There is small issue of using /usr/bin/ as working dir for running apps. You can't kill them with our kernel/daemon hotkeys, moreover random ppl using Windows will have an issue updating their binary. For now I think u should make entry in genimage.sh to move it to userspace, after 2.0.0 release I could change shipping method of RetroArch in upstream.

I can move it to /mnt during package generation

Some repos have additional data e.g. DAT files for mame cores, bios for gpsp, or required sounds for xrick etc... I think that this could be also included

**Comment by Apaczer** [comment]
> No, I checkout via git is slow, I'm trying to use tar.gz where it is posible

I see, we could setup a new fork of that repo, then generate all tarbals of different boxart, which could be then distributed&updated by separate runner there. At least that's what comes to me mind right now.

File: board/miyoo/scripts/genimage.sh [hunk 8]
```diff
@@ -14,6 +14,7 @@ cp -r board/miyoo/main "${BINARIES_DIR}"
 mkdir -p "${BINARIES_DIR}/gmenu2x"
```
Comment (Apaczer) [inline board/miyoo/scripts/genimage.sh]: why we make a dir here, can u recall?

File: package/libretro-cannonball/libretro-cannonball.mk [hunk 9]
```diff
@@ -19,8 +19,14 @@ endef
 
 define LIBRETRO_CANNONBALL_INSTALL_TARGET_CMDS
     mkdir -p "${BINARIES_DIR}/retroarch/cores"
+    mkdir -p "${BINARIES_DIR}/retroarch/system/cannonball/res"
 	$(INSTALL) -D $(@D)/cannonball_libretro.so \
 		${BINARIES_DIR}/retroarch/cores/cannonball_libretro.so
+        $(INSTALL) -D $(@D)/res/* \
+                ${BINARIES_DIR}/retroarch/system/cannonball/res
+        $(INSTALL) -D $(@D)/roms/* \
```
Comment (Apaczer) [inline package/libretro-cannonball/libretro-cannonball.mk]: for Cannoball you just need data package e.g. https://buildbot.libretro.com/assets/cores/Cannonball/CannonBall.zip - not sure about src of it.

Comment (Apaczer) [reply package/libretro-cannonball/libretro-cannonball.mk]: my bad, we can't run it without ROM

File: board/miyoo/scripts/genimage.sh (hunk 8 above)
Comment (tiopex) [reply board/miyoo/scripts/genimage.sh]: I think in the eariel version gmenu2x package produced only bin file, but I'm not sure. Now I think we can remove it

**Comment by Apaczer** [comment]
Moved all links to "cores" section, except for the RetroArch and RetroArch_History (last Played Session) which are in emus.

@tiopex , we're still missing scummvm core, there should be 70 in total cores at least.

**Comment by tiopex** [comment]
> Moved all links to "cores" section, except for the RetroArch and RetroArch_History (last Played Session) which are in emus.
> 
> @tiopex , we're still missing scummvm core, there should be 70 in total cores at least.

in the nightly build indeed there is scummvm core but in the latest stable not https://buildbot.libretro.com/stable/1.16.0/dingux/miyoo-arm32/RetroArch.7z, weird, or maybe it's not working correctly

anyway, I've added scummvm core to the build

**Comment by Apaczer** [comment]
I noticed that even bulk nightly packages have not all cores (missing at least dozen). With the scummvm I'm not sure if it's the buildbot, because indeed it doesn't work from current src or it just plays really slow (I've seen smth like that alread). They recently streamlined scummvm/scummvm with libretro and there were major changes.

**Comment by tiopex** [comment]
scumm it's not working for me, I suggest to exclude it from defconfig, and wait till it will be available in official stable builds

**Comment by Apaczer** [comment]
Try switching to old mirror: https://github.com/libretro-mirrors/scummvm (that one works), until they fix upstream. I can't get any error logs from current scummvm, it just hungs loading.

**Comment by Apaczer** [comment]
Also, maybe let us just build all cores from origin/master and lock on specific commit only these cores which pressent heavy regression (mame2003-plus might be also the case), what u think?

**Comment by tiopex** [comment]
The best option is to set commit ids from official stable releases, but I cant find it.

**Comment by tiopex** [comment]
I asked libretro folks and they said that the core bundles that come with the stables are just whatever is the current build at the time the stable was built

**Comment by Apaczer** [comment]
Right, so it justs happend to failed build scummvm at that time.

File: package/libretro-prboom/libretro-prboom.mk [hunk 10]
```diff
@@ -21,9 +21,12 @@ define LIBRETRO_PRBOOM_BUILD_CMDS
 endef
 
 define LIBRETRO_PRBOOM_INSTALL_TARGET_CMDS
-    mkdir -p "${BINARIES_DIR}/retroarch/cores"
+	mkdir -p "${BINARIES_DIR}/retroarch/cores"
+	mkdir -p "${BINARIES_DIR}/retroarch/system/prboom"
 	$(INSTALL) -D $(@D)/prboom_libretro.so \
 		${BINARIES_DIR}/retroarch/cores/prboom_libretro.so
+	$(INSTALL) -D $(@D)/prboom.wad \
+		${BINARIES_DIR}/retroarch/system/prboom
```
Comment (tiopex) [inline package/libretro-prboom/libretro-prboom.mk:29]: There is no possibility to select wad file from hidden directory

Comment (Apaczer) [reply package/libretro-prboom/libretro-prboom.mk:29]: This is just a data ROM, it is self-loaded at runtime I believe. You need also real WAD, for e.g. https://github.com/freedoom/freedoom/releases

Comment (tiopex) [reply package/libretro-prboom/libretro-prboom.mk:29]: No, if you load this wad it starts shareware doom

Comment (Apaczer) [reply package/libretro-prboom/libretro-prboom.mk:29]: Interesting, it does not start for me, also that's what written in manual.

Comment (tiopex) [reply package/libretro-prboom/libretro-prboom.mk:29]: Indeed, maybe I messed something up.

Comment (Apaczer) [reply package/libretro-prboom/libretro-prboom.mk:29]: I think no harm leaving it in ./system (it silences coreinfo warning), but if you want to cp it somewhere else it's ok.

**Comment by tiopex** [comment]
@Apaczer is it ready to merge?

**Review by Apaczer — APPROVED** [review]
Yeah, nextly we might:

- add 1 homebrew ROM (from e.g. [libretro site](https://buildbot.libretro.com/assets/cores/)) for every core to test + put them in appropriate folder thus associate Link's dir selector 
- add custom hotkey's configuration (made at compiletime or fetched from board)
- modify a bit frontend src (some changes never reached upstream I wished)

For now let's merge :100:
